src/main/java/org/tugraz/sysds/runtime/data/TensorBlock.java

```diff
@@ -100,7 +104,47 @@ public boolean isEmpty() {
 
 	public abstract double get(int r, int c);
 
+	/**
+	 * Set a cell to the value given as an `Object`. The type is inferred by either the `schema` or `valueType`, depending
+	 * if the `TensorBlock` is a `BasicTensor` or `DataTensor`.
+	 * @param ix indexes in each dimension, starting with 0
+	 * @param v value to set
+	 */
 	public abstract void set(int[] ix, Object v);
 
+	/**
+	 * Set a cell in a 2-dimensional tensor.
+	 * @param r row of the cell
+	 * @param c column of the cell
+	 * @param v value to set
+	 */
 	public abstract void set(int r, int c, double v);
+
+	/**
+	 * Aggregate a unary operation on this tensor.
+	 * @param op the operation to apply
+	 * @param result the result tensor
+	 * @return the result tensor
+	 */
+	public abstract TensorBlock aggregateUnaryOperations(AggregateUnaryOperator op, TensorBlock result);
+
+	public abstract void incrementalAggregate(AggregateOperator aggOp, TensorBlock partialResult);
+
+	public abstract TensorBlock binaryOperations(BinaryOperator op, TensorBlock thatValue, TensorBlock result);
+
+	public static ValueType resultValueType(ValueType in1, ValueType in2) {
+		if (in1 == ValueType.STRING || in2 == ValueType.STRING)
+			return ValueType.STRING;
+		if (in1 == ValueType.FP64 || in2 == ValueType.FP64)
+			return ValueType.FP64;
+		if (in1 == ValueType.FP32 || in2 == ValueType.FP32)
+			return ValueType.FP32;
+		if (in1 == ValueType.INT64 || in2 == ValueType.INT64)
+			return ValueType.INT64;
+		if (in1 == ValueType.INT32 || in2 == ValueType.INT32)
+			return ValueType.INT32;
+		if (in1 == ValueType.BOOLEAN || in2 == ValueType.BOOLEAN)
+			return ValueType.INT64;
```

Comment:
Why INT64? Ignoring UNKNOWN for know shouldn't it always be boolean & boolean, so return unconditionally return BOOLEAN here? Also please move a check for UNKNOWN with proper exception handling up into the beginning of the method.

Reply:
INT64 because for example `sum` should not have boolean as output. But it is true that we might want to overthink this method in itself, maybe once more operations are implemented add an operation parameter?